tests/test_env.py

```diff
@@ -894,3 +895,46 @@ def test_open_file_expired_aws_credentials(monkeypatch, caplog, path_rgb_byte_ti
         with rasterio.env.Env():
             with rasterio.open(path_rgb_byte_tif) as dataset:
                 assert not dataset.closed
+
+
+@pytest.fixture
+def http_server(tmpdir):
+    import functools
+    import multiprocessing
+    import http.server
+    from . import rangehttpserver
+    PORT = 8000
+    Handler = functools.partial(rangehttpserver.RangeRequestHandler, directory=str(tmpdir))
+    httpd = http.server.HTTPServer(("", PORT), Handler)
+    p = multiprocessing.Process(target=httpd.serve_forever)
+    p.start()
+    yield
+    p.terminate()
+    p.join()
+
+
+@pytest.mark.xfail(reason="GDAL has cached the first failed request")
+def test_vsi_curl_failure_cache(tmpdir, http_server):
+    """First failed request was cached"""
+    with pytest.raises(RasterioIOError):
+        rasterio.open("/vsicurl/http://localhost:8000/red.tif")
+
+    shutil.copy("tests/data/red.tif", str(tmpdir))
+
+    with rasterio.open("/vsicurl/http://localhost:8000/red.tif") as src:
+        assert src.count == 3
+        assert (src.read(1) == 204).all()
+
+
+@requires_gdal3(reason="Cache clearing requires GDAL 3+")
+def test_vsi_curl_cache_clear(tmpdir, http_server):
+    """Clearing cache wipes out previous failure"""
+    with pytest.raises(RasterioIOError):
+        rasterio.open("/vsicurl/http://localhost:8000/red.tif")
+
+    shutil.copy("tests/data/red.tif", str(tmpdir))
+
+    with rasterio.Env(clear_vsicurl_cache=True):
```

Comment:
This options says "clear GDAL's vsicurl cache on __enter__" and the previous failed request is wiped from our process's memory.